src/Parser.elm

```diff
@@ -0,0 +1,374 @@
+
+        Nothing :: aybes ->
+            maybeChain aybes
+
+        (Just m) :: _ ->
+            Just m
+
+
+cleanUpData : List F.AudienceFolder -> List A.Audience -> Data
+cleanUpData folders audiences =
+    makeModel
+        { folders = folders
+        , audiences = audiences
+        , model = zeroDataSet
+        , unique = 0
+        , folderIds = Dict.empty
+        , audienceIds = Dict.empty
+        }
+
+
+type alias MakeModel =
+    { folders : List F.AudienceFolder
+    , audiences : List A.Audience
+    , model : Data
+    , unique : Int
+    , folderIds : Dict.Dict Int Int
+    , audienceIds : Dict.Dict Int Int
+    }
+
+
+zeroDataSet : Data
+zeroDataSet =
+    { parents = Dict.empty
+    , audiences = Dict.empty
+    , all = Dict.empty
+    }
+
+
+type alias Data =
+    { parents : Dict.Dict Int Int
+    , audiences : Dict.Dict Int A.AudienceType
+    , all : Dict.Dict Int String
+    }
+
+
+makeModel : MakeModel -> Data
+makeModel m =
+    case m.folders of
+        [] ->
+            case m.audiences of
+                [] ->
+                    m.model
+
+                a :: udiences ->
+                    addAudience { m | audiences = udiences } a
+
+        f :: olders ->
+            addFolder { m | folders = olders } f
+
+
+addAudience : MakeModel -> A.Audience -> Data
+addAudience m a =
+    case a.folder of
+        Nothing ->
+            addRootAudience m a
+
+        Just parentId ->
+            addSubAudience m a parentId
+
+
+addFolder : MakeModel -> F.AudienceFolder -> Data
+addFolder m f =
+    case f.parent of
+        Nothing ->
+            addRootFolder m f
+
+        Just parentId ->
+            addSubFolder m f parentId
+
+
+addSubFolder : MakeModel -> F.AudienceFolder -> Int -> Data
+addSubFolder m f parentId =
+    let
+        newId =
+            m.unique
+
+        ( newParentId, newUnique ) =
+            case Dict.get parentId m.folderIds of
+                Nothing ->
+                    ( m.unique + 1, m.unique + 2 )
+
+                Just exists ->
+                    ( exists, m.unique + 1 )
+
+        model =
+            m.model
+
+        newModel =
+            { model
+                | parents = Dict.insert newId newParentId model.parents
+                , all = Dict.insert newId f.name model.all
+            }
+    in
+    makeModel
+        { m
+            | model = newModel
+            , unique = newUnique
+            , folderIds =
+                Dict.insert
+                    f.id
+                    newId
+                    (Dict.insert parentId newParentId m.folderIds)
+        }
+
+
+addSubAudience : MakeModel -> A.Audience -> Int -> Data
+addSubAudience m a parentId =
+    let
+        newId =
+            m.unique
+
+        ( newParentId, newUnique ) =
+            case Dict.get parentId m.folderIds of
+                Nothing ->
+                    ( m.unique + 1, m.unique + 2 )
+
+                Just exists ->
+                    ( exists, m.unique + 1 )
```

Comment:
Re the previous comment near `MakeModel`: this for example is something I'd love to _not_ have to write. It could be circumvented using something like
```elm
let
  audiencesCount = List.length audiences
  foldersCount = List.length audienceFolders
  newAudienceIds = List.range 0 (audiencesCount - 1)
  newAudienceFolderIds = List.range audiencesCount (audiencesCount + foldersCount - 1)
in
...
```
and then just zipping them with the audiences list so that you have both the ID and the audience it should be applied to, in the same lambda.

But again, I think a cleaner solution would be to _not_ give new IDs to the audiences and folders.